cdc/sink/codec/canal_flat.go

```diff
@@ -0,0 +1,238 @@
+
+package codec
+
+import (
+	"encoding/json"
+	"github.com/pingcap/errors"
+	"github.com/pingcap/log"
+	"github.com/pingcap/ticdc/cdc/model"
+	cerrors "github.com/pingcap/ticdc/pkg/errors"
+	canal "github.com/pingcap/ticdc/proto/canal"
+	"go.uber.org/zap"
+)
+
+// CanalFlatEventBatchEncoder encodes Canal flat messages in JSON format
+type CanalFlatEventBatchEncoder struct {
+	builder       *canalEntryBuilder
+	unresolvedBuf []*canalFlatMessage
+	resolvedBuf   []*canalFlatMessage
+}
+
+// NewCanalFlatEventBatchEncoder creates a new CanalFlatEventBatchEncoder
+func NewCanalFlatEventBatchEncoder() EventBatchEncoder {
+	return &CanalFlatEventBatchEncoder{
+		builder:       NewCanalEntryBuilder(),
+		unresolvedBuf: make([]*canalFlatMessage, 0),
+		resolvedBuf:   make([]*canalFlatMessage, 0),
+	}
+}
+
+type canalFlatMessage struct {
+	ID            int64                    `json:"id"`
+	Schema        string                   `json:"database"`
+	Table         string                   `json:"table"`
+	PKNames       []string                 `json:"pkNames"`
+	IsDDL         bool                     `json:"isDdl"`
+	EventType     string                   `json:"type"`
+	ExecutionTime int64                    `json:"es"`
+	BuildTime     uint64                   `json:"ts"`
+	Query         string                   `json:"sql"`
+	SQLType       map[string]int32         `json:"sqlType"`
+	MySQLType     map[string]string        `json:"mysqlType"`
+	Data          []map[string]interface{} `json:"data"`
+	Old           []map[string]interface{} `json:"old"`
+	tikvTs        uint64
+}
+
+func (c *CanalFlatEventBatchEncoder) newFlatMessageForDML(e *model.RowChangedEvent) (*canalFlatMessage, error) {
+	eventType := convertRowEventType(e)
+	header := c.builder.buildHeader(e.CommitTs, e.Table.Schema, e.Table.Table, eventType, 1)
+	rowData, err := c.builder.buildRowData(e)
+	if err != nil {
+		return nil, cerrors.WrapError(cerrors.ErrCanalEncodeFailed, err)
+	}
+
+	pkCols := e.PrimaryKeyColumns()
+	pkNames := make([]string, len(pkCols))
+	for i := range pkNames {
+		pkNames[i] = pkCols[i].Name
+	}
+
+	var nonTrivialRow []*canal.Column
+	if e.IsDelete() {
+		nonTrivialRow = rowData.BeforeColumns
+	} else {
+		nonTrivialRow = rowData.AfterColumns
+	}
+
+	sqlType := make(map[string]int32, len(nonTrivialRow))
+	mysqlType := make(map[string]string, len(nonTrivialRow))
+	for i := range nonTrivialRow {
+		sqlType[nonTrivialRow[i].Name] = nonTrivialRow[i].SqlType
+		mysqlType[nonTrivialRow[i].Name] = nonTrivialRow[i].MysqlType
+	}
+
+	var (
+		data    map[string]interface{}
+		oldData map[string]interface{}
+	)
+
+	if len(rowData.BeforeColumns) > 0 {
+		oldData = make(map[string]interface{}, len(rowData.BeforeColumns))
+		for i := range rowData.BeforeColumns {
+			if !rowData.BeforeColumns[i].GetIsNull() {
+				oldData[rowData.BeforeColumns[i].Name] = rowData.BeforeColumns[i].Value
+			} else {
+				oldData[rowData.BeforeColumns[i].Name] = nil
+			}
```

Comment:
If the column is nil, what should be output in Canal rule, `null` or empty string(`""`)

If using string is valid in data and oldData, we can save some CPU cost of interface convert 

The Canal Column provides the following method 
https://github.com/pingcap/ticdc/blob/fc58c002bb6efec14b8c4e88ab8325e7e43e1c06/proto/canal/EntryProtocol.pb.go#L622-L627

Reply:
Empty string in not supported in Canal Json because there would be no place in the Json to indicate whether the value is really null or just an empty string if the column is a string type. Canal Protobuf supports this because of the `IsNullPresent` field.